### PR TITLE
Added "PHP Type Providers" article for PhpStorm topic

### PR DESCRIPTION
**Pull request by i582**
This article briefly describes the type system in PhpStorm and the implementation of the type provider.

### CODE REVIEW

**Comment by YannCebron** [comment]
Thanks, will be reviewed shortly

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+Type inference in PhpStorm is built on top of type providers, each of which is responsible for inferring the types of
+specific PSI elements.
+For example, `com.jetbrains.php.lang.psi.resolve.types.PhpArrayAccessTP` is responsible for inferring the types of expressions like `$arr[10]`.
+There are dozens of such providers, and they all work one after another to provide type information when needed.
+
+All providers inherit from `com.jetbrains.php.lang.psi.resolve.types.PhpTypeProvider4`, which is registered in the `com.jetbrains.php.typeProvider4` extension point.
+
+## Types in PhpStorm
+
+The first stage of type inference in PhpStorm takes place at the indexing stage, at this moment PhpStorm don’t
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:14]: To make this fully correct I'd also specify that during indexing it's possible to use content of local current file to obtain such information. For some cases it can represent effectively complete information for the project, for instance declaration and usages of private method in the current file.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+know anything about the functions that are in the code, don’t know what types they return, don’t know the classes
+and what methods are called, all this imposes restrictions of how types work in PhpStorm.
+
+Because of this, types in PhpStorm are divided into two types:
+
+- Complete types
+- Incomplete types
+
+Internally, types are stored in the index as strings, these strings can contain one type or several if it is a [Union Type](#union-types).
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:23]: I'm not sure I fully understood this, but objects of class `PhpType` are used to do any operations and access on PHP types. In current implementation this class contains collection of strings and each string should contain one and only one effective type, complete or incomplete.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+<?php
+
+function foo(): string { return "Hello World!"; }
+```
+
+Here, since the `foo()` function has a type hint, the return type is also Complete.
+
+### Incomplete types
+
+Incomplete types are types that require additional resolving and contain additional information for accurate type inference.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:47]: Incomplete types are types that require additional information from other files in the project besides containing file.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+Because of this, types in PhpStorm are divided into two types:
+
+- Complete types
+- Incomplete types
+
+Internally, types are stored in the index as strings, these strings can contain one type or several if it is a [Union Type](#union-types).
+
+### Complete types
+
+Complete types are types that are already known exactly, for example, in the following code:
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:27]: Complete types are types that are known exactly based on the content of the containing file

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+function foo(): Foo { return new Foo(); }
+
+foo()->f();
+```
+
+Here PhpStorm must somehow express the type of the expression `foo()` during indexing, however,
+during indexing, PhpStorm cannot refer to the definition of the `foo()` function to return a specific type.
+
+So PhpStorm creates an Incomplete type into which it writes all the information it might need to resolve the type.
+In this case, the name of the function and its arguments, so PhpStorm will create a new Incomplete type `#琁\foo.`.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:66]: Correct signature will be `#F\foo`

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+#### Incomplete Types Structure
+
+At the very beginning there is a `#` character, which is a marker that the type is Incomplete.
+
+It is followed by the type provider's unique key.
+According to it, you can understand that this type was created by the provider of the desired type.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:73]: It's actually created not for user understanding but rather to be able to know from serialised type what provider completion phase to invoke. Client code has no control over this.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+The rest of the line is the encoded information.
+In the example above, it's the fully qualified name and a dot.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:76]: Should be updated after providing correct signature.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+The created Incomplete type will be processed further when the indexing process is over.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:78]: I'd specify explicitly that `PhpTypeProvider4#complete` will be invoked for this type at current stage

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+Here PhpStorm must somehow express the type of the expression `foo()` during indexing, however,
+during indexing, PhpStorm cannot refer to the definition of the `foo()` function to return a specific type.
+
+So PhpStorm creates an Incomplete type into which it writes all the information it might need to resolve the type.
+In this case, the name of the function and its arguments, so PhpStorm will create a new Incomplete type `#琁\foo.`.
+
+#### Incomplete Types Structure
+
+At the very beginning there is a `#` character, which is a marker that the type is Incomplete.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:70]: I'd compose this paragraph to specify explicitly that `PhpTypeProvider4#getType` will be invoked first during indexing is supposed to provide either additional complete type or incomplete type that will be passed to the same type provider `PhpTypeProvider4#complete` during global mode to infer complete type.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+if (rand(0, 10) > 5) {
+  $a = "Hello World!".
+}
+
+$a; // (1)
+```
+
+In (1), the variable `$a` will be of type `int|string` because PhpStorm can't deduce exactly which branch the execution will take.
+
+PhpStorm stores these types by separating each of the individual types with a `|` character.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:116]: It's not correct, `|` used only for representation purposes. Multiple types are stored as multiple strings inside `PhpType`

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+
+2. `getType()` is a method that returns the type of the expression for the given element.
+   This method is called at the indexing stage and therefore its implementation **cannot access** any information
+   from the index and **must** rely only on local information.
+   If you need some information, then pack the required data into a string and return an Incomplete type based on this string.
+
+3. `complete()` is a method that resolves an Incomplete type into a Complete type.
+   All strings of Incomplete types are sequentially passed to this method, the method should return
+   a Complete type for them.
+   Note that Incomplete types from other type providers also come to this method,
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:201]: Actually this should not happen, can you please provide the scenario to reproduce.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+   so you need to check that the incoming string belongs to your provider.
+
+4. `getBySignature()` is a method with which you can provide additional elements or references.
+
+You can also overload the `emptyResultIsComplete()` method, which indicates whether the `null` returned from
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:206]: It's unclear what "is a valid result" means in this context. What actually happens is that PhpStorm won't add `mixed` for the unresolved type in case this method returns `true`.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+the `complete()` method is a valid result.
+For example, if you index the type `int`, then the result returned from `complete()` will be `null` and it is a valid result.
+
+**Examples:**
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:210]: I'd not use there type providers as an idiomatic examples since they're pretty hard to follow and have non-standard implementations. Let's leave `PhpUnitFiledInitializedInSetUpMethodsTP` as the only example implementation

**Review by wbars — CHANGES_REQUESTED** [review]
Thank you for contribution! Left some notes.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 1 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 2 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 3 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 5 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 6 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 7 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 10 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 11 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: It looks like I made a mistake here, rechecked and did not find such situations 

Fixed

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 12 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Added.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 13 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+
+## PhpType
+
+PhpStorm uses the `com.jetbrains.php.lang.psi.resolve.types.PhpType` class to work with types.
+
+To add types to it, use `add()`, which can take either another `PhpType` or a `string`.
+
+To check that a type is Complete, use `isComplete()`.
+
+To resolve the Incomplete type, use `global()`.
```
Comment (wbars) [inline topics/products/phpstorm/php_open_api_php_type_providers.md:134]: It's worth to repeat the context IMO, please add

"This method shouldn't be used during indexing, namely inside `PhpTypeProvider4#getType`"

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+For example, `com.jetbrains.php.lang.psi.resolve.types.PhpArrayAccessTP` is responsible for inferring the types of expressions like `$arr[10]`.
+There are dozens of such providers, and they all work one after another to provide type information when needed.
+
+All providers inherit from `com.jetbrains.php.lang.psi.resolve.types.PhpTypeProvider4`, which is registered in the `com.jetbrains.php.typeProvider4` extension point.
+
+## Types in PhpStorm
+
+The first phase of type inference takes place at the indexing stage.
+At this phase, PhpStorm calls `PhpTypeProvider4.getType()` on each type provider.
+PhpStorm only has access to local information from the current file and cannot use information from others.
```
[the remark this answers is not in the document]
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+```
+<?php
+
+$a = 100;
+```
+
+The type of the expression that's assigned to the `$a` variable is Complete type `int`, since PhpStorm can definitely infer that a numeric literal is of type `int`.
+
+```
+<?php
```
[the remark this answers is not in the document]
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md:37]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md (hunk 14 above)
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md:134]: Fixed.

**Review by wbars — APPROVED** [review]
Thank you for the contribution, this is valuable addition.

**Review by YannCebron — CHANGES_REQUESTED** [review]

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+import com.jetbrains.php.lang.psi.elements.*;
+import com.jetbrains.php.lang.psi.elements.impl.PhpClassImpl;
+import com.jetbrains.php.phpunit.PhpUnitUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.Set;
+
+public class PhpUnitFiledInitializedInSetUpMethodsTP implements PhpTypeProvider4 {
```
[the remark this answers is not in the document]
Comment (YannCebron) [reply topics/products/phpstorm/php_open_api_php_type_providers.md:237]: OK

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+<path>main.php</path>:
+
+```php
+<?php
+
+$a = foo();
+echo $a;
+```
+
+In the `main.php` file, we call the `foo()`, which is defined in another `foo.php` file.
```
[the remark this answers is not in the document]
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Fixed.

File: topics/products/phpstorm/php_open_api_php_type_providers.md [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+3. `complete()` resolves an Incomplete type into a Complete type.
+   All strings of Incomplete types are sequentially passed to this method, it should return a Complete type for them.
+
+4. `getBySignature()` provides additional elements or references.
+
+You can also override the `emptyResultIsComplete()`, which indicates whether the `null` returned from the `complete()` is a valid result, which means that PhpStorm won't add the `mixed` to the resulting type.
+
+## Example Implementation
+
+The goal of this example is to provide types for filed references assigned in `setUp` method if containing class is PHPUnit one.
```
[the remark this answers is not in the document]
Comment (i582) [reply topics/products/phpstorm/php_open_api_php_type_providers.md]: Yep, fixed.
